File: cpp/StftPitchShift/Cepster.cpp
#include <StftPitchShift/Cepster.h>

Cepster::Cepster(const std::shared_ptr<FFT> fft, const double quefrency, const double samplerate) :
  fft(fft),
  quefrency(static_cast<size_t>(quefrency * samplerate))
{
}

void Cepster::lifter(const std::vector<std::complex<float>>& dft, std::vector<float>& envelope)
{
  spectrum.resize(dft.size());
  cepstrum.resize((dft.size() - 1) * 2);
  envelope.resize(dft.size());

  for (size_t i = 0; i < spectrum.size(); ++i)
  {
    spectrum[i] = std::log10(dft[i].real());
  }

  fft->ifft(spectrum, cepstrum);
  lowpass(cepstrum, quefrency);
  fft->fft(cepstrum, spectrum);

  for (size_t i = 0; i < spectrum.size(); ++i)
  {
    envelope[i] = std::pow(10.0f, spectrum[i].real());
  }
}

void Cepster::lowpass(std::vector<float>& cepstrum, const size_t quefrency)
{
  for (size_t i = 1; i < std::min(quefrency, cepstrum.size()); ++i)
  {
    cepstrum[i] *= 2;
  }

  for (size_t i = quefrency + 1; i < cepstrum.size(); ++i)
  {
    cepstrum[i] = 0;
  }
}

File: python/stftpitchshift/cepster.py
import numpy as np


def lowpass(cepstrum, quefrency):

    cepstrum[1:quefrency] *= 2
    cepstrum[quefrency+1:] = 0

    return cepstrum


def lifter(frames, quefrency):

    envelopes = np.zeros(frames.shape)

    for i, frame in enumerate(frames):

        with np.errstate(divide='ignore', invalid='ignore'):
            spectrum = np.log10(np.real(frame))

        cepstrum = np.fft.irfft(spectrum, norm='forward')
        cepstrum = lowpass(cepstrum, quefrency)
        cepstrum = np.fft.rfft(cepstrum, norm='forward')

        envelopes[i] = np.power(10, np.real(cepstrum))

    return envelopes
